implementation/irq-closures.tex

```diff
@@ -0,0 +1,350 @@
+In order to implement the pattern above we need to get the globally defined interrupt handler \texttt{ADC0\_IRQHandler} to call a closure created in the main function.
+So lets first consider what a closure is.
+A closure is the \textbf{environment} it closes over, in \autoref{lst:irq:local} the \texttt{ch} variable, and a \textbf{function pointer}, that simple.
+See \autoref{sec:back:rust:closures} for additional details of closures.
+
+An initial naive idea of registering the function pointer from the clousre in the interrupt vector will not work as the function nees to be called with the environment in the closure.
+Therfore we need to store the whole closure in a datastructure and ensure that it is called by the interrupt handler.
+First we try to store the closure in a global datastructure which the handler can access as implemented in \autoref{lst:global-closure}.
+
+\begin{listing}[H]
+  \begin{minted}{rust}
+static mut CLOSURE: Option<Fn()> = None;
+  \end{minted}
+  \caption{Storing the closure globally}
+  \label{lst:global-closure}
+\end{listing}
+
+In \autoref{lst:global-closure} we introduce the \texttt{Fn} trait.
+This is a trait which a represent closure in Rust.
+Wrapping the closure in an \texttt{Option} is done to enable the handler to be undefined.
+Compiling this will yield an error due to the fact that the compiler cannot statically determine the size of the closure and therefore does not know how much memory to allocate.
+This problem occures quite common in programming and the default solution is to put the value behind a pointer as shown in \autoref{lst:global-pointer-to-closure}
+
+\begin{listing}[H]
+  \begin{minted}{rust}
+static mut CLOSURE: Option<Box<Fn()>> = None;
+  \end{minted}
+  \caption{Storing a pointer to the closure globally}
+  \label{lst:global-pointer-to-closure}
+\end{listing}
+
+This introduces a Box which yields an indirection, storing the closure on the heap.
+Unfortunatly this does not make the program compile.
+It is due to the fact alluded to earlier saying that global variables cannot have a type which does not have a constant-expression initializer.
+The Box type is one of these types which are disallowed as global variables.
+
+To get around this issue we need to use raw pointers.
+We substitue the \texttt{Box} pointer with a raw pointer to the closure as shown in \autoref{lst:raw-global-pointer-to-closure}.
+
+\begin{listing}[H]
+  \begin{minted}{rust}
+static mut CLOSURE: Option<*const Fn()> = None;
+  \end{minted}
+  \caption{Storing a pointer to the closure globally}
+  \label{lst:raw-global-pointer-to-closure}
+\end{listing}
+
+Now we've managed to store a closure in a statically allocated location.
+\autoref{lst:pub-sub} shows a safe abstraction to register and dispatch events with this handle.
+
+\begin{listing}[H]
+  \begin{minted}{rust}
+// Registers an interrupt handler
+fn register(f: Box<Fn()>) {
```

Comment:
Denne koden tar opp mer plass enn vi har på en side, den burde gjøres kortere.

Kan f.eks. endres til noe ala koden under. Den må nok endres for å få den til å kompilere, men den er ca 15 linjer kortere som gjør at den passer bedre i rapporten.
```rust
// Registers an interrupt handler
fn register(f: Box<Fn()>) {
  unsafe {
    // Deallocate old handler if existing
    if let Some(old) = CLOSURE {
        // Remove the global reference
        CLOSURE = None;
        // Return the ownership of the pointer
        // to a managed box, this will ensure
        // that the contents is deallocated when
        // the box variable goes out of scope
        let box = boxed::from_raw(old);
    }
    // Consume the Box pointer and
    // return a raw pointer to the closure
    let raw = boxed::into_raw(f);
    // Save the closure pointer in the global reference
    CLOSURE = Some(raw);
  }
}
// Dispatch a event by calling the closure if it is registered
fn dispatch() {
  // The Closure is stored in a global variable mutable
  // So the access to this variable is unsafe
  unsafe {
    // Unwrap the closure value
    if let Some(ref func) = CLOSURE {
        // The closure is stored behind a pointer which must
        // be dereferenced, it is called with is environment
        // by invoking the function: call
        (*func).call(())
    }
  }
}

```